plasmapy/diagnostics/thomson.py

```diff
@@ -141,177 +360,587 @@ def spectral_density(
+
+    """
+
+    required_settings = {
+        "probe_wavelength",
+        "probe_vec",
+        "scatter_vec",
+        "ions",
+    }
+
+    if missing_settings := required_settings - set(settings):
+        raise ValueError(
+            f"The following required settings were not provided in the "
+            f"'settings' argument: {missing_settings}"
         )
 
-    icontr = np.zeros([ifract.size, w.size], dtype=np.complex128) * u.s / u.rad
-    for m in range(ifract.size):
-        icontr[m, :] = ifract[m] * (
-            2
-            * np.sqrt(np.pi)
-            * ion_z[m]
-            / k
-            / vTi[m]
-            * np.power(np.abs(np.sum(chiE, axis=0) / epsilon), 2)
-            * np.exp(-xi[m, :] ** 2)
+    required_params = {"n"}
+    if missing_params := required_params - set(params):
+        raise ValueError(
+            f"The following required parameters were not provided in the "
+            f"'params': {missing_params}"
         )
 
-    # Recast as real: imaginary part is already zero
-    Skw = np.real(np.sum(econtr, axis=0) + np.sum(icontr, axis=0))
+    # **********************
+    # Count number of populations
+    # **********************
+    if "efract_0" not in params:
+        params.add("efract_0", value=1.0, vary=False)
+
+    if "ifract_0" not in params:
+        params.add("ifract_0", value=1.0, vary=False)
+
+    num_e = _count_populations_in_params(params, "efract")
+    num_i = _count_populations_in_params(params, "ifract")
+
+    # **********************
+    # Required settings and parameters per population
+    # **********************
+    for p, nums in zip(["T_e", "T_i"], [num_e, num_i]):
+        for num in range(nums):
+            key = p + "_" + str(num)
+            if key not in params:
+                raise ValueError(
+                    f"{p} was not provided in kwarg 'parameters', but is required."
+                )
+
+    # **************
+    # ions
+    # **************
+
+    ions = settings["ions"]
+    # Condition ions
+    # If a single value is provided, turn into a particle list
+    if isinstance(ions, ParticleList):
+        pass
+    elif isinstance(ions, str):
+        ions = ParticleList([Particle(ions)])
+    # If a list is provided, ensure all values are Particles, then convert
+    # to a ParticleList
+    elif isinstance(ions, list):
+        for ii, ion in enumerate(ions):
+            if isinstance(ion, Particle):
+                continue
+            ions[ii] = Particle(ion)
+        ions = ParticleList(ions)
+    else:
+        raise ValueError(
+            "The type of object provided to the ``ions`` keyword "
+            f"is not supported: {type(ions)}"
+        )
 
-    return np.mean(alpha), Skw
+    # Validate ions
+    if len(ions) == 0:
+        raise ValueError("At least one ion species needs to be defined.")
+
+    try:
+        if sum([ion.charge_number <= 0 for ion in ions]):
```

Comment:
```suggestion
        if sum(ion.charge_number <= 0 for ion in ions):
```
I think we could use a generator here.